drivers/gpio/gpio_sx1509b.c

```diff
@@ -104,112 +106,107 @@ static int gpio_sx1509b_config(struct device *dev, int access_op, u32_t pin,
-			pins->pull_up |= BIT(pin);
-		} else {
-			pins->pull_up &= ~BIT(pin);
-		}
-		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
-			pins->pull_down |= BIT(pin);
-		} else {
-			pins->pull_down &= ~BIT(pin);
-		}
-		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_NORMAL) {
-			pins->pull_up &= ~BIT(pin);
-			pins->pull_down &= ~BIT(pin);
-		}
-		if (flags & GPIO_DS_DISCONNECT_HIGH) {
-			pins->open_drain |= BIT(pin);
-		} else {
-			pins->open_drain &= ~BIT(pin);
-		}
-		if (flags & GPIO_POL_INV) {
-			pins->polarity |= BIT(pin);
-		} else {
-			pins->polarity &= ~BIT(pin);
-		}
-		break;
-	case GPIO_ACCESS_BY_PORT:
-		if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
-			pins->dir = 0xffff;
-			pins->input_disable = 0x0000;
-		} else {
-			pins->dir = 0x0000;
-			pins->input_disable = 0xffff;
-		}
-		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
-			pins->pull_up = 0xffff;
-		} else {
-			pins->pull_up = 0x0000;
-		}
-		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
-			pins->pull_down = 0xffff;
-		} else {
-			pins->pull_down = 0x0000;
-		}
-		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_NORMAL) {
-			pins->pull_up = 0x0000;
-			pins->pull_down = 0x0000;
-		}
-		if (flags & GPIO_DS_DISCONNECT_HIGH) {
-			pins->open_drain = 0xffff;
-		} else {
-			pins->open_drain = 0x0000;
-		}
-		if (flags & GPIO_POL_INV) {
-			pins->polarity = 0xffff;
-		} else {
-			pins->polarity = 0x0000;
-		}
-		break;
+	if ((flags & GPIO_OPEN_MASK) == GPIO_OPEN_DRAIN) {
+		pins->open_drain |= bit;
+	} else {
+		pins->open_drain &= ~bit;
+	}
+
+	switch (flags & GPIO_PUD_MASK) {
 	default:
-		ret = -ENOTSUP;
+		ret = -EINVAL;
 		goto out;
+	case GPIO_PUD_NORMAL:
+		pins->pull_up &= ~bit;
+		pins->pull_down &= ~bit;
+		break;
+	case GPIO_PUD_PULL_UP:
+		pins->pull_up |= bit;
+		pins->pull_down &= ~bit;
+		break;
+	case GPIO_PUD_PULL_DOWN:
+		pins->pull_up &= ~bit;
+		pins->pull_down |= bit;
+		break;
 	}
 
-	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
-				    SX1509B_REG_DIR, pins->dir);
-	if (ret)
-		goto out;
+	if (flags & GPIO_IO_IN_ENABLED) {
+		pins->input_disable &= ~bit;
+	} else {
+		pins->input_disable |= bit;
+	}
 
-	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
-				    SX1509B_REG_INPUT_DISABLE,
-				    pins->input_disable);
-	if (ret)
-		goto out;
+	if (flags & GPIO_IO_OUT_ENABLED) {
+		pins->dir &= ~bit;
+	} else {
+		pins->dir |= bit;
+	}
 
-	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
-				    SX1509B_REG_PULL_UP,
-				    pins->pull_up);
-	if (ret)
-		goto out;
+	if (flags & GPIO_IO_INIT_ENABLED) {
+		data_first = true;
+		if (flags & GPIO_IO_INIT_HIGH) {
+			pins->data |= bit;
+		} else {
+			pins->data &= ~bit;
+		}
+	}
 
-	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
-				    SX1509B_REG_PULL_DOWN,
-				    pins->pull_down);
-	if (ret)
-		goto out;
+	if (flags & GPIO_POL_INV) {
+		pins->polarity |= bit;
+	} else {
+		pins->polarity &= !bit;
```

Comment:
This won't work as expected. Once the polarity is inversed for a given bit, there will be no way to restore the normal operation for it, as `!bit` will be always `0`. Change to `~bit`.